Repository pattern applied for the persistence layer

RoxanaTanaselea opened the pull request:
I implemented the repository pattern for the persistence layer. The purpose is to have a clean separation between the domain layer and the persistence layer and to make the data origin transparent to the client. "Client objects construct query specifications declaratively and submit them to Repository for satisfaction. Objects can be added to and removed from the Repository, as they can from a simple collection of objects, and the mapping code encapsulated by the Repository will carry out the appropriate operations behind the scenes" (Martin Fowler)
The way I did it :
I defined the Repository Pattern for Station entity
I implemented a basic Repository with Sqlite 
I used a data mapper layer to isolate domain objects from details of the database access code
I did Unit Tests for Sqlite repository implentation

Commit:
created the station model pojos

Commit:
updated some gradle dependencies

Commit:
created repository and specification interfaces
represent the base definition of a Repository and a Specification

Commit:
created the database contract and helper

Commit:
created dataMapper interface
implemented by StationContentValueDataMapper, StationCursorDataMapper, StationRealmDataMapper
used for implementig the DataMapper pattern

Commit:
created interface SqlSpecification that inherits Specification
used for SqlRepository implementation

Commit:
created StationSqlRepository
implements Repository for Sqlite (for Station table)

Commit:
UnitTests for StationSqlRepository

eduardb commented on app/build.gradle:
```diff
@@ -12,6 +12,7 @@ buildscript {
 apply plugin: 'com.android.application'
 apply plugin: 'me.tatarka.retrolambda'
 apply plugin: 'com.neenbedankt.android-apt'
+apply plugin: 'realm-android'
```
I don't see `realm` used in the project, why have you applied the plugin?

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/api/ClujBikeService.java:
```diff
@@ -7,6 +7,6 @@
 public interface ClujBikeService {
 
-    @GET("Station/Read")
+    @GET("StationApiEntity/Read")
```
I believe that this got here by mistake after a refactoring, the API endpoint shouldn't have changed.

eduardb replied:
Also, I just double checked on the website, and the method should be `POST`, not `GET`. Could you correct, please?

eduardb commented on build.gradle:
```diff
@@ -25,9 +26,9 @@ task clean(type: Delete) {
 ext {
-    compileSdkVersion = 24
-    targetSdkVersion = 24
-    buildToolsVersion = "24.0.3"
```
Please also update `circle.yml` with the corresponding changes

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/api/model/StationApiEntity.java:
```diff
@@ -14,7 +14,7 @@
 
     @SerializedName("OcuppiedSpots")
-    public int ocuppiedSpots;
+    public int occupiedSpots;
```
😂 👍

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/DatabaseHelper.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
+        db.execSQL(SqlDataBaseContract.Station.DROP_TABLE);
```
It's a bad practice, in general, to drop a table when upgrading. Migration is possible 99% of the time. But anyway, you're at the first version of the DB, there's no upgrade happening yet, so YAGNI.

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/Repository.java:
```diff
@@ -0,0 +1,17 @@
+import rx.Completable;
+import rx.Observable;
+
+interface Repository<T> {
```
👍

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/SqlDataBaseContract.java:
```diff
@@ -0,0 +1,56 @@
+
+import android.provider.BaseColumns;
+
+public final class SqlDataBaseContract {
```
Isn't "database" a single word?

eduardb replied:
Why not `abstract`? Or better yet, an `interface`?

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/SqlDataBaseContract.java:
```diff
@@ -0,0 +1,56 @@
+    public static final int DATA_BASE_VERSION = 1;
+    public static final String DATA_BASE_NAME = "clujbike.db";
+
+    public static final String TEXT_TYPE = "TEXT";
```
You should better add some spaces around the strings, otherwise your SQL queries will be incorrect.

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/SqlDataBaseContract.java:
```diff
@@ -0,0 +1,56 @@
+
+    private SqlDataBaseContract(){}
+
+    public static abstract class Station implements BaseColumns {
```
This also looks as a good candidate for an `interface`

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/SqlDataBaseContract.java:
```diff
@@ -0,0 +1,56 @@
+        public static final String COLUMN_STATUS = "status";
+        public static final String COLUMN_STATUS_TYPE = "statusType";
+        public static final String COLUMN_LATITUDE = "latitude";
+        public static final String COLUMN_LONGITUDE = "";
```
Whoops, empty strings?

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/StationSqlRepository.java:
```diff
@@ -0,0 +1,93 @@
+
+public class StationSqlRepository implements Repository<StationEntity> {
+
+    static final String STATION_TABLE_NAME = Station.TABLE_NAME;
```
Do you think that this new constant brings any value? Isn't it redundant?

eduardb replied:
bump

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/StationSqlRepository.java:
```diff
@@ -0,0 +1,93 @@
+                                DataMapper<StationEntity, ContentValues> contentValueDataMapper,
+                                DataMapper<Cursor, StationEntity> cursorDataMapper) {
+        this.databaseHelper = databaseHelper;
+
```
Why the blank line? 💃

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/StationSqlRepository.java:
```diff
@@ -0,0 +1,93 @@
+        this.cursorDataMapper = cursorDataMapper;
+    }
+
+    private void closeTransactionWithoutSuccess(SQLiteDatabase db) {
```
Please move methods after their first usage

eduardb replied:
bump

RoxanaTanaselea replied:
solved in commit 1c15c8a

eduardb commented on build.gradle:
```diff
@@ -5,7 +5,8 @@ buildscript {
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.2.0'
+        classpath 'com.android.tools.build:gradle:2.3.0-alpha2'
```
I am a bit reluctant in using alpha versions of the build tools gradle plugin

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/mapper/DataMapper.java:
```diff
@@ -0,0 +1,5 @@
+package com.deveddy.clujbike.data.repository.mapper;
+
+public interface DataMapper<T, V> {
+    V mapFrom(T response);
```
Is the argument really a `response`?

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/StationSqlRepository.java:
```diff
@@ -0,0 +1,93 @@
+        return Observable.just(item)
+                .map(contentValueDataMapper::mapFrom)
+                .doOnNext(contentValues
+                        -> db.update(STATION_TABLE_NAME,
```
Wouldn't it be better to move this line next to the argument? The `->` kinda gets lost here...

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/StationSqlRepository.java:
```diff
@@ -0,0 +1,93 @@
+                    db.delete(STATION_TABLE_NAME,
+                            sqlSpecification.whereClause(),
+                            sqlSpecification.whereArgs());
+                    db.close();
```
Wouldn't be worth to move this call in a `doOnComplete()`, in case that `delete()` throws (and to be consisted with the rest of this class)?

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/mapper/station/StationRealmDataMapper.java:
```diff
@@ -0,0 +1,24 @@
+import com.deveddy.clujbike.data.repository.mapper.DataMapper;
+import com.deveddy.clujbike.data.repository.models.StationRealm;
+
+public class StationRealmDataMapper implements DataMapper<StationApiEntity, StationRealm> {
```
So you have classes related to Realm, but no repository. How about removing these classes for now, and adding Realm support in a follow-up PR?

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/StationSqlRepository.java:
```diff
@@ -0,0 +1,93 @@
+        SQLiteDatabase db = databaseHelper.getReadableDatabase();
+        SqlSpecification sqlSpecification = ((SqlSpecification) specification);
+        return Observable
+                .fromCallable(() -> db.rawQuery(sqlSpecification.toString(), sqlSpecification.whereArgs()))
```
Why not using `db.query()` instead, and `sqlSpecification.whereClause()`?

eduardb commented on app/src/test/java/com/deveddy/clujbike/data/repository/StationSqlRepositoryTest.java:
```diff
@@ -0,0 +1,246 @@
+
+        verify(contentValuesDataMapper).mapFrom(any(StationEntity.class));
+        testSubscriber.assertError(error);
+        //// TODO: 24.11.2016  
```
What's with this `TODO`?

eduardb commented on app/src/test/java/com/deveddy/clujbike/data/repository/StationSqlRepositoryTest.java:
```diff
@@ -0,0 +1,246 @@
+                .observeOn(Schedulers.immediate())
+                .subscribe(testSubscriber);
+
+        testSubscriber.assertError(error);
```
Probably you should also verify that `db.close()` was called

eduardb replied:
bump

RoxanaTanaselea replied:
solved in 	56ebe0e

Commit:
removed realm support and related classes

Commit:
edited getStations() call
changed endpoint
changed http method (from GET to POST)

Commit:
updated circle.yml with corresponding changes

Commit:
eliminated onUpgrade() implementation method from DataBaseHelper

Commit:
changed SqlDatabaseContract
transformed the class into an Interface
renamed interface from SqlDataBaseContract to SqlDatabaseContract
corrected  the sql query syntax
transformed Station into an Interface

Commit:
updated StationSqlRepository
small syntax changes
corrected query and remove methods

Commit:
changed build tools version to a stable one

Commit:
minor syntax changes

Commit:
updated DataMapper mapFrom method's parameter name

Commit:
modified SqlSpecification declaration
removed toString() method

Commit:
updated StationSqlRepositoryTest class

Commit:
did Instrumentation tests for sqlStationMappers
tested StationContentValueDataMapper
tested StationCursorDataMapper

Commit:
Fix file name issue

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/api/model/StationApiEntity.java:
```diff
@@ -5,16 +5,16 @@
     @SerializedName("Address")
     public String address;
 
-    @SerializedName("OcuppiedSpots")
```
Serialized name shouldn't have changed, the mistake is in their API

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/api/ClujBikeService.java:
```diff
@@ -2,11 +2,13 @@
 public interface ClujBikeService {
 
-    @GET("Station/Read")
+    @FormUrlEncoded
```
There are no fields, so it shouldn't be encoded

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/api/model/StationApiEntity.java:
```diff
@@ -5,16 +5,16 @@
 import java.util.Date;
 import java.util.List;
```
Unused import

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/SqlDatabaseContract.java:
```diff
@@ -0,0 +1,49 @@
+        String COLUMN_CUSTOM_IS_VALID = "customIsValid";
+
+        String CREATE_TABLE = "CREATE_TABLE " + TABLE_NAME + " (" +
+                _ID + " INTEGER PRIMARY KEY" + COMMA +
```
Why not extracting `"PRIMARY KEY"` in a constant as well?

RoxanaTanaselea replied:
Because, for the time being, it is used only once. It would be redundant

eduardb commented on app/src/androidTest/java/com/deveddy/clujbike/data/repository/mapper/station/StationCursorDataMapperTest.java:
```diff
@@ -0,0 +1,67 @@
+        String[] columns = new String[]{"_id", "name", "address", "occupiedSpots",
+                "emptySpots", "maximumNumberOfBikes", "lastSyncDate", "idStatus", "status",
+                "statusType", "latitude", "longitude", "isValid", "customIsValid"};
+        MatrixCursor cursor = new MatrixCursor(columns);
```
Nice 👍

eduardb commented on app/src/androidTest/java/com/deveddy/clujbike/data/repository/mapper/station/StationCursorDataMapperTest.java:
```diff
@@ -0,0 +1,67 @@
+import static junit.framework.Assert.assertEquals;
+
+@RunWith(AndroidJUnit4.class)
+public class StationCursorDataMapperTest {
```
What happens when `cursor` is `null`? How about when it doesn't have any rows?

RoxanaTanaselea replied:
solved in 	7947ad6

eduardb commented on app/src/androidTest/java/com/deveddy/clujbike/data/repository/mapper/station/StationContentValueDataMapperTest.java:
```diff
@@ -0,0 +1,67 @@
+
+
+@RunWith(AndroidJUnit4.class)
+public class StationContentValueDataMapperTest {
```
What happens when the `ContentValues` object is null or contains invalid data?

RoxanaTanaselea replied:
solved in 	7947ad6

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/DatabaseHelper.java:
```diff
@@ -0,0 +1,21 @@
+public class DatabaseHelper extends SQLiteOpenHelper {
+
+    public DatabaseHelper(Context context) {
+        super(context, SqlDatabaseContract.DATA_BASE_NAME, null, SqlDatabaseContract.DATA_BASE_VERSION);
```
If `database` is one word, shouldn't the constant's name be `DATABASE`? :trollface:

RoxanaTanaselea replied:
true, my bad

eduardb commented on app/src/main/java/com/deveddy/clujbike/data/repository/mapper/DataMapper.java:
```diff
@@ -0,0 +1,5 @@
+package com.deveddy.clujbike.data.repository.mapper;
+
+public interface DataMapper<T, V> {
+    V mapFrom(T t);
```
What if we would change the names here? I think that the following would be simpler and cleaner:
```java
public interface Mapper<F, T> {
    T from(F from);
}
```
`T` would stand for "to" and `F` would stand for "from".
What do you think?

RoxanaTanaselea replied:
It would simplify the writing. I will change it

eduardb reviewed (changes requested):
Getting there :P Try not to miss comments.

Commit:
move methods after their first usage in StationSqlRepository

Commit:
correct stationApi key spelling

Commit:
erase service method encoding

Commit:
refactored DataMapper to Mapper
simplify Mapper interface declaration syntax

Commit:
small syntax refactor
refactor DATA_BASE_NAME to DATABASE_NAME

Commit:
add new tests to mapperTests
-in StationContentValueMapperTest cover the case where statioEntity is incorrectly built and the case where stationEntity is empty
-in StationCursorMapperTest cover the case where cursor has no rows and the case where cursor is null

Commit:
fix StationSqlRepository tests
- close database when an Error is thrown
- change methods tested that did not handled the onError case

eduardb commented on app/src/androidTest/java/com/deveddy/clujbike/data/repository/mapper/station/StationCursorMapperTest.java:
```diff
@@ -0,0 +1,81 @@
+        assertEquals(station.customValid(), getBooleanValue(cursor, "customIsValid"));
+    }
+
+    private Cursor givenCursor(boolean noRows) {
```
Wouldn't `hasRows` make more sense than `noRows`? It will avoid the negation later, and it would make more sense where you call it (`givenCursor(true)` instead of `givenCursor(false)` where you want a cursor with rows)

eduardb commented on build.gradle:
```diff
@@ -5,7 +5,7 @@ buildscript {
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.2.0'
+        classpath 'com.android.tools.build:gradle:2.2.2'
```
`2.2.3` just came out, just sayin' 😛

Commit:
rename variable in StationSqlRepository

Commit:
updated circle.yml

Commit:
update tools.build.gradle version to 2.2.3

Commit:
small refactoring in StationCursorMapperTest

Commit:
Revert "updated circle.yml"

This reverts commit 0a2d0343548a669a8b0e587098a55f1916241772.
